### ISSUE TITLE
Regular expression / String problem that was previously satisfiable now reports unsat

### ISSUE THREAD

**Issue by jawline** [Z3Prover/z3#1028]
Hi, The query listed below was satisfiable with previous versions of z3 (And is still sat on http://rise4fun.com/z3) however testing with the newest version of z3 now reports unsat. Is this just a bug or is there something wrong with my query?

Thanks for any help.

```
(declare-fun |1 Fill 0| () String)
(declare-fun TestC0 () String)
(declare-fun |1 Fill 2| () String)
(declare-fun |1 Fill 1| () String)
(assert (= |1 Fill 0| "xxx"))
(assert (let ((a!1 (re.++ (re.* (re.union (re.range "\x00" "\x09") (re.range "\v" "y")))
                  (str.to.re "xxx"))))
(let ((a!2 (re.++ a!1
                  (re.* (re.union (re.range "\x00" "\x09") (re.range "\v" "y"))))))
  (str.in.re TestC0 a!2))))
(assert (let ((a!1 (re.++ (re.* (re.union (re.range "\x00" "\x09") (re.range "\v" "y")))
                  (str.to.re "xxx"))))
(let ((a!2 (re.++ a!1
                  (re.* (re.union (re.range "\x00" "\x09") (re.range "\v" "y"))))))
  (or (not (str.in.re TestC0 a!2))
      (= TestC0 (str.++ |1 Fill 1| (str.++ "xxx" |1 Fill 2|)))))))

(check-sat)
(get-model)
```

**Comment by PierreVDL** [comment]
I suspect it has the same rootcause as issue #998

**Comment by NikolajBjorner** [comment]
@mtrberzi: the code that rewrites characters to strings would create the empty string from the 0 character. This basically inherits C semantics that strings are 0 terminated lists of characters.

**Comment by mtrberzi** [comment]
It looks like this is fixed by ceec81de as the instance appears to be SAT in the latest commit. I'll run my tests to make sure this doesn't affect anything in Z3str3.

**Comment by jawline** [comment]
Yep - Solves now. Thanks!
